src/backend/distributed/commands/alter_table.c

```diff
@@ -0,0 +1,1303 @@
+			TableConversionParameters cascadeParam = {
+				.relationId = colocatedTableId,
+				.shardCountIsNull = con->shardCountIsNull,
+				.shardCount = con->shardCount,
+				.colocateWith = qualifiedRelationName,
+				.cascadeToColocated = CASCADE_TO_COLOCATED_NO_ALREADY_CASCADED,
+				.spiAlreadyConnected = true
+			};
+			TableConversionReturn *colocatedReturn = con->function(&cascadeParam);
+			foreignKeyCommands = list_concat(foreignKeyCommands,
+											 colocatedReturn->foreignKeyCommands);
+		}
+	}
+
+	/* recreate foreign keys */
+	TableConversionReturn *ret = NULL;
+	if (con->conversionType == ALTER_DISTRIBUTED_TABLE)
+	{
+		if (con->cascadeToColocated != CASCADE_TO_COLOCATED_NO_ALREADY_CASCADED)
+		{
+			char *foreignKeyCommand = NULL;
+			foreach_ptr(foreignKeyCommand, foreignKeyCommands)
+			{
+				spiResult = SPI_execute(foreignKeyCommand, false, 0);
+				if (spiResult != SPI_OK_UTILITY)
+				{
+					ereport(ERROR, (errmsg("could not run SPI query")));
+				}
+			}
+		}
+		else
+		{
+			ret = malloc(sizeof(TableConversionReturn));
+			ret->foreignKeyCommands = foreignKeyCommands;
+		}
+	}
+
+	if (!con->spiAlreadyConnected)
+	{
+		spiResult = SPI_finish();
+		if (spiResult != SPI_OK_FINISH)
+		{
+			ereport(ERROR, (errmsg("could not finish SPI connection")));
+		}
+	}
+
+	return ret;
+}
+
+
+/*
+ * EnsureTableNotReferencing checks if the table has a reference to another
+ * table and errors if it is.
+ */
+void
+EnsureTableNotReferencing(Oid relationId)
+{
+	if (TableReferencing(relationId))
+	{
+		ereport(ERROR, (errmsg("cannot complete operation "
+							   "because table has a foreign key")));
+	}
+}
+
+
+/*
+ * EnsureTableNotReferenced checks if the table is referenced by another
+ * table and errors if it is.
+ */
+void
+EnsureTableNotReferenced(Oid relationId)
+{
+	if (TableReferenced(relationId))
+	{
+		ereport(ERROR, (errmsg("cannot complete operation "
+							   "because a foreign key references to table")));
+	}
+}
+
+
+/*
+ * EnsureTableNotForeign checks if the table is a foreign table and errors
+ * if it is.
+ */
+void
+EnsureTableNotForeign(Oid relationId)
+{
+	char relationKind = get_rel_relkind(relationId);
+	if (relationKind == RELKIND_FOREIGN_TABLE)
+	{
+		ereport(ERROR, (errmsg("cannot complete operation "
+							   "because it is a foreign table")));
+	}
+}
+
+
+/*
+ * EnsureTableNotPartition checks if the table is a partition of another
+ * table and errors if it is.
+ */
+void
+EnsureTableNotPartition(Oid relationId)
+{
+	if (PartitionTable(relationId))
+	{
+		Oid parentRelationId = PartitionParentOid(relationId);
+		char *parentRelationName = get_rel_name(parentRelationId);
+		ereport(ERROR, (errmsg("cannot complete operation "
+							   "because table is a partition"),
+						errhint("the parent table is \"%s\"",
+								parentRelationName)));
+	}
+}
+
+
+TableConversionState *
+CreateTableConversion(TableConversionParameters *params)
+{
+	TableConversionState *con = malloc(sizeof(TableConversionState));
```

Comment:
we should never ever call malloc in postgres code, it will create a memory leak

use palloc instead, that makes sure that the memory is released by postgres at the end of the statement/transaction